Fix an assertion failure in tls_common.c

mattcaswell opened the pull request:
When we clear the record layer, we better make sure we clear all relevant fields, otherwise we can get ourselves into an unexpected state.

Fixes #23255

mattcaswell commented:
Still draft because I need a test.

nhorman commented on ssl/ssl_lib.c:
```diff
@@ -560,17 +560,35 @@ void OPENSSL_VPROC_FUNC(void) {}
 
-    /* We try and reset both record layers even if one fails */
+    /* Clear any buffered records we no longer need */
+    while (s->rlayer.curr_rec < s->rlayer.num_recs)
```
should ret be included in the while condition here so that you break out of the loop if any record release fails? i.e.:
```
int ret = 1;

while (ret == 1 && s->rlayer.curr_rec < s->rlayer.num_recs)
   ret &= ssl_release_record(...)

if (ret == 0)
   return 0;

mattcaswell replied:
Actually I wondered about that myself. But in the end I thought that this really should not fail - and if it does then shouldn't we make a best effort to release everything, and then only fail at the end?

mattcaswell commented:
I have added a test to this. The test revealed further problems, which I have also fixed. The fixes then caused other tests to fail in the QUIC code - due to some latent issues that were revealed by my changes.

This is ready for review.

hlandau commented on test/sslapitest.c:
```diff
@@ -7000,22 +7000,45 @@ static int test_key_update_local_in_read(int tst)
+/*
+ * Test clearing a connection via SSL_clear(), or resetting it via
+ * SSL_set_connect_state()/SSL_set_accept_state()
+ * Test 0: SSL_set_conect_state, TLSv1.3
```
nit: connect

hlandau reviewed (changes requested):
Looks good

t8m commented on test/sslapitest.c:
```diff
@@ -7054,6 +7125,7 @@ static int test_ssl_clear(int idx)
     return testresult;
 }
 
+
```
Spurious empty line added

t8m commented on ssl/statem/statem_lib.c:
```diff
@@ -1981,12 +1981,13 @@ int ssl_version_supported(const SSL_CONNECTION *s, int version,
          ++vent) {
         if (vent->cmeth != NULL
                 && ssl_version_cmp(s, version, vent->version) == 0
-                && ssl_method_error(s, vent->cmeth()) == 0
+                && ssl_method_error(s, s->server ? vent->smeth()
```
Should this be also conditioned on s->server?

mattcaswell replied:
Not sure I understand the question? It *is* conditioned on `s->server`?

t8m replied:
And them I would just factor out this ternary to a temporary SSL_METHOD * variable.

t8m reviewed (changes requested):
One question and one nit

mattcaswell commented:
Updated to address the outstanding comment.

t8m commented:
Hmm, CI looks relevant.

mattcaswell commented:
> Hmm, CI looks relevant.

Fixed. Hopefully.

mattcaswell commented:
@t8m - can you reconfirm?

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed to master/3.2. There was a trivial conflict when cherry-picking to 3.2 which I resolved during merge.
